include/utilities/mdframe/mdframe.hpp

```diff
@@ -0,0 +1,273 @@
+     * Otherwise, returns boost::none.
+     * 
+     * @param name Name of the variable.
+     * @return boost::optional<variable>
+     */
+    variable& operator[](const std::string& name) noexcept
+    {
+        return this->get_variable(name);
+    }
+
+    /**
+     * Check if a variable exists.
+     * 
+     * @param name Name of the variable.
+     * @return true if the variable exists in this mdframe.
+     * @return false if the variables does **not** exist in this mdframe.
+     */
+    bool has_variable(const std::string& name) const noexcept
+    {
+       return this->find_variable(name) != this->m_variables.end();
+    }
+
+    /**
+     * Add a (vector) variable definition to this mdframe.
+     *
+     * @note This member function is constrained by @c{Args}
+     *       being constructible to a std::initializer_list of std::string
+     *       (i.e. a list of std::string).
+     * 
+     * @tparam Arg list of std::string representing dimension names.
+     * @param name Name of the variable.
+     * @param dimensions Names of the dimensions this variable spans.
+     * @return mdframe&
+     */
+    template<typename T, types::enable_if_supports<T, bool> = true>
+    mdframe& add_variable(const std::string& name, std::initializer_list<std::string> dimensions)
+    {
+        std::vector<dimension> references;
+        references.reserve(dimensions.size());
+
+        for (const auto& d : dimensions) {
+            auto dopt = this->get_dimension(d);
+            if (dopt == boost::none) {
+                throw std::runtime_error("not a dimension");
+            }
+            references.push_back(dopt.get());
+        }
+
+        this->m_variables[name] = variable::make<T>(name, references);
+    
+        return *this;
+    }
+
+
+    // ------------------------------------------------------------------------
+    // Output Member Functions
+    // ------------------------------------------------------------------------
+
+    /**
+     * Write this mdframe to a CSV file.
+     *
+     * The outputted CSV is a projected (or normalized)
+     * table spanning all possible dimensions, similar
+     * to the description of tidy data by Hadley Wickham (2014):
+     * https://www.jstatsoft.org/article/view/v059i10
+     *
+     * @note The dimensions are not outputted to the CSV,
+     *       since they only form the basis for the mdframe's
+     *       dimensions, and do not inherently store any information.
+     * 
+     * @param path File path to output CSV
+     * @param basis Dimension to span CSV over, defaults to "time"
+     * @param header Should the CSV header be included?
+     */
+    void to_csv(const std::string& path, bool header = true) const;
+
+#if NETCDF_ACTIVE
+    /**
+     * Write this mdframe to a NetCDF file.
+     * 
+     * @param path File path to the output NetCDF
+     */
+    void to_netcdf(const std::string& path) const;
+#endif
```

Comment:
For purposes of things like ccache, it's slightly preferable to not conditionalize the declaration, and either leave the definition out, so linking fails, or have it error at runtime.

Not necessary to change for this PR, though.

Reply:
Okay, that makes sense -- I wasn't sure if there were any issues with this. Either in this PR or a future one, I'll move the conditional to the source file so that if NetCDF isn't available, then this function will throw a runtime error.